### ISSUE TITLE
Thoughts on best strategy to apply semi-supervised learning to my dataset

### ISSUE THREAD

**Issue by otmarjr** [MaartenGr/BERTopic#241]
First of all, I'd like to thank you for the great effort put in this project. It's a really cool project and extremelly easy to use. Congratulations on the nice work done here! 👍 

I'd like to have some help with my problem, which seems quite challenging and after running Bertopic, I ended up a little confused.

I'm working on classifier based on tickets in our ITSM system, where I'd like to automatically label issues that should be forwarded to the correct support team. I started working with the main IT service we support: a legal application for managing cases through the court system. I must add that I work for the Labour Court here in Brazil. 

One of our IT team leads classified around 2.000 tickets that should be used to guide the semi-supervised process. However, after running Bertopic, I cannot see any meaningful connection between the topics discovered by Bertopic and those previously classified by our IT specialist.

Currently, my labeled dataset looks like this:

```
df['TOPIC_ID'] = pd.Categorical(df.ID_CLASSE).codes
df['TOPIC_ID'].value_counts(ascending=False)
```
output:
12    622
31    300
28    230
20    192
11     77
...
3       1
40      1
1       1

I tried to limit my analysis to classes containing at least 100 samples. This reduced my dataset to only 4 classes and  1344 records:

```
df2 = df.groupby("TOPIC_ID").filter(lambda r: len(r) > MIN_REGS_CLASSE).copy()
labels = df2.TOPIC_ID.to_list()
docs = df2.TEXTO.to_list()
topic_model = BERTopic(language="portuguese")
topic_model.fit(docs, y=labels)
```

I looped through the topic_model.topics and compared this topic to the class originaly labeled by the IT specialist by keeping a dictionary for each original label and a list for corresponding topics:

```
classes_vs_tids = {}

for tid in topic_model.topics:
    if tid == -1:
        continue
        
    id_classes = []
    indices_representativos = []
    
    docs = topic_model.get_representative_docs(tid)
    for doc in docs:
        reg = df.loc[df.TEXTO == doc]
        if len(reg) > 0:
            id_classe = reg.ID_CLASSE.values[0]
            if id_classe in classes_vs_tids:
                classes_vs_tids[id_classe] = classes_vs_tids[id_classe] | {tid} 
            else: 
                classes_vs_tids[id_classe] = {tid} 
```

And then, I compared Bertmodel's topics and the original labels:

```
for idc in classes_vs_tids:
    total_tids = len(classes_vs_tids[idc])
        print(f"Class {id2classe[idc]} corresponds to {total_tids} topics:  {classes_vs_tids[idc]}")
    for tid in classes_vs_tids[idc]:
        print(f"\tRelevant terms for TID {tid}: {topic_model.get_topic(tid)}")
```

Even after restricting to only 4 classes, I was very surprised to discover that only one of those four classes correspond to just one topic. Two classes correspond to 7 topics and the last one to 13 topics!

Knowing that those tickets originally were logged by several different people, I guess that some variability is to be expected. However, I'm curious to know if there's any setting or strategy to make Bertopic lean torwards my original classification.

**Comment by MaartenGr** [comment]
Thank you for the extensive description, that definitely gives me a good idea of what you are trying to accomplish. 

First a few notes. Seeing as you have partially labeled data it might be interesting to also use the unlabeled data in BERTopic. Those, you can give a -1 label to indicate that they do not have a label. That will help in more accurately creating lower-dimensional representations. Moreover, if you have more documents, you will most likely create a larger number of topics that will be of higher quality. 

On to the question at hand. BERTopic often creates a large number of topics. Since we are working with document embeddings that contain a lot of information regarding the documents, HDBSCAN is able to create many topics. In practice, this means that fine-grained topics will be created that could be merged together (depending on your definition of the abstractness of a topic). 

So when you have some labels to your dataset, there is a good chance that within those labels, BERTopic will find sub-topics that can and perhaps should be separated. Especially with topic modeling, there is a degree of subjectivity involved when creating (sub-)topics. 

Secondly, the semi-supervised approach of BERTopic only nudges the creation of topics towards the labels if it makes sense. This means that the labels you supply to BERTopic are merely a suggestion to the model. In practice, if it can find more fine-grained topics or the labeling does not make sense, it is highly unlikely it will actually cluster them together. 

Although I am not familiar with your data, I believe that a combination of the above is happening. Seeing as the labels are generated by a domain expert, I will assume that labels are mostly accurate. Given that assumption, it is highly likely that simply more fine-grained topics are being created for each labeled topic. That means that if one class corresponds to 7 topics, those 7 topics might be related to each other and similarly related to the one class. 

Also, I am curious if you have checked whether the labeled topic matches with the related topics. For example, if you have 1 labeled topic that corresponds to 4 topics, do those 4 topics make sense? Moreover, what is the distribution of those 4 topics? If one topic is significantly more frequent than the other 3, then that is a small error of margin (which might actually be attributed to the labeling process). 

Hopefully, this helps a bit. If not, please let me know!

**Comment by otmarjr** [comment]
Thanks for the quick response, Marteen!

I skimmed through some examples and I perceived that the topics are indeed very fined grained. For example: there's a class related to problems with the injunctions management screens and taks. There are 622 examples in this class which have been divided in fairly evenly matched topics summing up 562 records in those topics:

![image](https://user-images.githubusercontent.com/110718/133784539-3b1d8198-d543-4ee2-8647-717b2fdf7ed7.png)

After analyzing some samples, I noted that Bertopic clustered by some extra details like:

- Who is trying to update the injunction - Judicial Officer or lawyer
- Different error messages in the very same screen
- Problems with the underlying form: blank fields, mispelled information or corrections to involved parties' data.

This is indeed very fine, because it really detected different type of problems for just one of the software features!

However, I'm not sure if the domain expert will be able to manage and deal with so many more labels. His IT team is comprised of people who, in general, can deal with problems in a per-feature basis. If I ask him to label and analyze even finer details, I guess his classification process will be even slower. I'm trying to use Bertopic to 'give him a hand' and accelerate the classification process.

**Comment by MaartenGr** [comment]
Let me start off by saying thank you for the extensive and clear response, this definitely helps in constructing the next steps. 

From your description, it seems that the main issue is then indeed that the topics are simply too fine-grained for your use case. I can imagine that in practice, it would be difficult to label those finer details. Moreover, you can argue that such fine-grained topics are actually not meaningful from a classification perspective. 

If I am correct, the solution then seems to be relatively straightforward, reduce the number of topics to reduce fine-grained topics. There are several ways in approaching that:

First, since you are working with domain experts, you can try to merge topics manually. For example, construct your own clusters of topics that are talking about similar topics. You can do this by reading through all topics yourself and simply map topics together. For example, if topic 1 is related to topics 87 and 34, then every time you predict topic 87 or 34, you instead return topic 1. This has a number of downsides though as the representation of topic 1 only takes documents in topic 1 into account. Moreover, this can actually be quite a bit of work to do manually. However, sometimes it is necessary if significant domain knowledge is needed. 

Second, you can try to reduce the number of topics before training the model by specifying `nr_topics=200`. When setting the number of topics manually, it tries to merge small topics with larger topics in order to remove those fine-grained topics. The benefit of doing this is that the model will focus on the largest topics and iteratively merge smaller topics. Also, you can try `nr_topics="auto"` to only merge topics that are very similar instead of forcefully merging topics. 
Similarly, you can reduce the number of topics after training the model by running `new_topics, new_probs = topic_model.reduce_topics(docs, topics, probabilities=probs, nr_topics=30)`. 

Third, and this the preferred method and one that I would highly advise, is increasing the `min_topic_size` parameter to decrease the number of generated topics. The larger the minimum topic size, the less fine-grained topics will be created. Setting this to `60` instead of the default `10` will definitely help to reduce the number of generated topics. This method works quite well without the need to artificially merge topics.

**Comment by otmarjr** [comment]
Hey Marteen!

I tried this third idea and then I got much better results by specifying min_topic_size=100 (in line with my initial criteria of learning only for larger classes). I still have to do some debug for one of the classes which ended up split in 4 topics. Nonetheless, it seems much more reasonable now.

I'd like to ask one final direction: some issues contain hardly any useful data for my classifier, because it mentions similar issues. For example:

> Dear colleagues, good afternoon.
> The problem reported on ticket REQ-XYZ3213 that we tought was solved last week is happening again. Could you check?
> Kind regards

I srtarted labeling those in a separate class, which I may try to deal via regexes when I detect that it's referencing another ticket. However, I'm unsure if I should create a different class for them or try to bin them in that special -1 topic id, which may contain other very different records for further analysis (which, in general, start popping after new features are released in newer versions).

**Comment by MaartenGr** [comment]
Glad to hear that the results are getting better! 

Generally, if there are documents that are unrelated I would simply put them in the -1 topic. That topic contains only outliers and technically should not be considered as a topic but a bunch of documents that are unrelated to each other. 

Let me know if you have any other questions.

**Comment by otmarjr** [comment]
Got it!

You've made it very clear, Marteen, thank you very much.  I'm closing the issue, ok?
